tmt/steps/prepare/shell.py

```diff
@@ -172,33 +193,94 @@ def go(
-            if not self.is_dry_run:
-                prepare_wrapper_path.chmod(0o755)
-            guest.push(
-                source=prepare_wrapper_path,
-                destination=prepare_wrapper_path,
-                options=TransferOptions(protect_args=True, preserve_perms=True, chmod=0o755),
-            )
-            command: ShellScript
+
+            script_name = f'{self.name} / script #{script_index}'
+
+            script_record_dirpath = self.phase_workdir / f'script-{script_index}' / guest.safe_name
+            script_log_filepath = script_record_dirpath / 'output.txt'
+
+            script_log_filepath.parent.mkdir(parents=True, exist_ok=True)
+            script_log_filepath.touch()
+
+            script_environment = environment.copy()
+
+            pull_options = DEFAULT_PULL_OPTIONS.copy()
+            pull_options.exclude.append(str(script_log_filepath))
+
+            script = tmt.utils.ShellScript(f'{tmt.utils.SHELL_OPTIONS}; {script}')
+
             if guest.become and not guest.facts.is_superuser:
-                command = tmt.utils.ShellScript(f'sudo -E {prepare_wrapper_path}')
+                command = tmt.utils.ShellScript(f'sudo -E {script}')
             else:
-                command = tmt.utils.ShellScript(f'{prepare_wrapper_path}')
-            guest.execute(
-                command=command,
-                cwd=worktree,
-                env=environment,
-                sourced_files=[self.step.plan.plan_source_script],
+                command = tmt.utils.ShellScript(f'{script}')
+
+            def _invoke_script(
+                command: ShellScript,
+                environment: tmt.utils.Environment,
+            ) -> tmt.utils.CommandOutput:
+                guest.push(source=self.phase_workdir)
+
+                return guest.execute(
+                    command=command,
+                    cwd=worktree,
+                    env=environment,
+                    sourced_files=[self.step.plan.plan_source_script],
+                )
+
+            output, error, timer = Stopwatch.measure(_invoke_script, command, script_environment)
+
+            if isinstance(error, tmt.utils.RunError):
+                self._post_action_pull(
+                    guest=guest,
+                    path=self.phase_workdir,
+                    pull_options=pull_options,
+                    exceptions=outcome.exceptions,
+                )
+
+                return self._save_failed_run_outcome(
+                    log_filepath=script_log_filepath,
+                    label=script_name,
+                    timer=timer,
+                    command=command,
+                    exception=error,
+                    outcome=outcome,
+                )
+
+            if isinstance(error, Exception):
+                return self._save_error_outcome(
+                    label=script_name, exception=error, outcome=outcome
+                )
```

Comment:
Didn't you make it in the `write_report` so that both `RunError` and general `Exception` can be handled in the same function?